### PR TITLE
Add demo for image classification models evaluation

### PR DESCRIPTION
**Pull request by wojtuss**
This patch adds a demo C-API application for image classification models evaluation.

The `infer_image_classification` application works fine with ResNet50, SE-ResNeXt50 and MobileNet-v1 models containing accuracy and topk layers so accuracy values could be calculated. It is also assumed the ImageNet dataset is available.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -195,6 +195,7 @@ include(external/gtest)     # download, build, install gtest
 include(external/protobuf)  # download, build, install protobuf
 include(external/python)    # download, build, install python
 include(external/openblas)  # download, build, install openblas
+include(external/opencv)    # download, build, install opencv
```
Comment (luotao1) [inline CMakeLists.txt:198]: - please do not add `external/opencv`, could you provide the preprocess data such as `vis_demo.cc` do?
- You can `run.sh`, and the download package which has both model and data.
- And we don't need `image_reader.cc`

File: paddle/fluid/inference/api/demo_ci/CMakeLists.txt [hunk 2]
```diff
@@ -71,6 +71,7 @@ if (NOT WIN32)
 link_directories("${PADDLE_LIB}/third_party/install/snappy/lib")
```
Comment (luotao1) [inline paddle/fluid/inference/api/demo_ci/CMakeLists.txt:71]: please do not change this file

File: paddle/fluid/inference/api/demo_ci/build.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
+set -x
```
Comment (luotao1) [inline paddle/fluid/inference/api/demo_ci/build.sh:1]: please not add `build.sh`, you can update `run.sh`

File: paddle/fluid/inference/api/demo_ci/infer_image_classification.cc [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+// dimensions of imagenet images are assumed as default:
+DEFINE_int32(resize_size, 256,
+             "Images are resized to make smaller side have length resize_size "
+             "while keeping aspect ratio.");
+DEFINE_int32(crop_size, 224,
+             "Resized images are cropped to crop_size x crop_size.");
+DEFINE_int32(channels, 3, "Width of the image.");
+
+namespace {
+class Timer {
```
Comment (luotao1) [inline paddle/fluid/inference/api/demo_ci/infer_image_classification.cc:52]: put `Timer` into `utils.h`

File: paddle/fluid/inference/api/demo_ci/run.sh [hunk 5]
```diff
@@ -5,7 +5,7 @@ TEST_GPU_CPU=$3 # test both GPU/CPU mode or only CPU mode
 DATA_DIR=$4 # dataset
 TENSORRT_INCLUDE_DIR=$5 # TensorRT header file dir, defalut to /usr/local/TensorRT/include
 TENSORRT_LIB_DIR=$6 # TensorRT lib file dir, default to /usr/local/TensorRT/lib
-inference_install_dir=${PADDLE_ROOT}/build/fluid_inference_install_dir
+inference_install_dir=${PADDLE_ROOT}/build/fluid_install_dir
```
Comment (luotao1) [inline paddle/fluid/inference/api/demo_ci/run.sh:8]: line 8 should not be updated.

File: paddle/fluid/inference/api/demo_ci/run.sh [hunk 6]
```diff
@@ -41,81 +41,89 @@ function download() {
...
 :<<D
   # -----simple_on_word2vec-----
-  rm -rf *
-  cmake .. -DPADDLE_LIB=${inference_install_dir} \
-    -DWITH_MKL=$TURN_ON_MKL \
-    -DDEMO_NAME=simple_on_word2vec \
-    -DWITH_GPU=$TEST_GPU_CPU \
-    -DWITH_STATIC_LIB=$WITH_STATIC_LIB
-  make -j
+  build_demo simple_on_word2vec
```
Comment (luotao1) [inline paddle/fluid/inference/api/demo_ci/run.sh:68]: You don't use the latest code of this file.

File: paddle/fluid/inference/api/demo_ci/infer_image_classification.cc [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+  config.specify_input_name = false;
+  config.SetCpuMathLibraryNumThreads(FLAGS_num_threads);
+  if (FLAGS_with_mkldnn) config.EnableMKLDNN();
+
+  // remove all passes so that we can add them in correct order
+  for (int i = config.pass_builder()->AllPasses().size() - 1; i >= 0; i--)
+    config.pass_builder()->DeletePass(i);
+
+  // add passes
+  config.pass_builder()->AppendPass("infer_clean_graph_pass");
```
Comment (luotao1) [inline paddle/fluid/inference/api/demo_ci/infer_image_classification.cc:196]: do not add pass manually.

**Comment by wojtuss** [comment]
@luotao1 
We would like to merge the application to `PaddlePaddle/Paddle` repository for several reasons:
- we would like to have a C-API analogs of python scripts from `PaddlePaddle/models` repository for inference, evaluation (with accuracy) and performance benchmarking of image classification models,
- as with the python scripts, we would like to test the models using real datasets, like ImageNet, (hence we use the OpenCV library, the other demo apps do not use real images),
- for the benchmarking results to be valuable, the code we use for benchmarking should be public, so that everyone could verify them.

To me, the `PaddlePaddle/models` repository would suit best for these purposes, but I understand your objection as the repo does not have CI enabled.

Please, give us an advice on how we should proceed to enable our application.

**Comment by luotao1** [comment]
@wojtuss 

Thanks for your explanation!
> for inference, evaluation (with accuracy) and performance benchmarking

How about add a `analyzer_xxx_tester` to do evaluation and performance benchmarking at first? 
- Unit-test is much easiler than demo_ci, since unit-tests already contains `profile_mkldnn`, `compare_mkldnn`, `multi-threads profile`.
- You can check the evaluation accuracy in unit-test as well.

>  real datasets, like ImageNet

How about preprocess the real datasets in your sides, and give us a `data.txt`, since we couldn't include `opencv` library in our third-party.

> for the benchmarking results to be valuable, the code we use for benchmarking should be public, so that everyone could verify them

Agree with you! We are planning to public a serial of inference benchmarking and check them every day now. How about add to unit-test at first, and wait us a few days?

**Comment by wojtuss** [comment]
@luotao1 ,
Thank you for your answer. I appreciate your help.
We would prefer to keep the application in the standalone form, so we decided to place it and the other similar applications on a public server hosted by Intel. This will let us keep the applications aligned with the other testing and benchmarking tools we have.

**Pull request closed without merging**